apps/game_backend/lib/game_backend/users.ex

```diff
@@ -42,29 +45,92 @@ defmodule GameBackend.Users do
       iex> get_user("9483ae81-f3e8-4050-acea-13940d47d8ed")
       nil
   """
-  def get_user(id), do: Repo.get(User, id) |> preload()
+  def get_user(id) do
+    user = Repo.get(User, id) |> preload()
+    if user, do: {:ok, user}, else: {:error, :not_found}
+  end
 
   @doc """
   Gets a user by their username.
 
-  Returns nil if no user is found.
+  Returns {:error, :not_found} if no user is found.
 
   ## Examples
 
       iex> get_user_by_username("some_user")
-      %User{}
+      {:ok, %User{}}
 
       iex> get_user_by_username("non_existing_user")
-      nil
+      {:error, :not_found}
   """
-  def get_user_by_username(username), do: Repo.get_by(User, username: username) |> preload()
+  def get_user_by_username(username) do
+    user = Repo.get_by(User, username: username) |> preload()
+    if user, do: {:ok, user}, else: {:error, :not_found}
+  end
 
   def update_experience(user, params),
     do:
       user
       |> User.experience_changeset(params)
       |> Repo.update()
 
+  @doc """
+  Increments a user's level and apply the level's rewards.
+  If it was the last level in the campaign, increments the campaign number and sets the level number to 1.
+  """
+  def advance_level(user_id, campaign_id) do
+    with {:campaign_data, {:ok, campaign_progress}} <-
+           {:campaign_data, retrieve_campaign_progress_data(user_id, campaign_id)},
+         {:next_level, {next_campaign_id, next_level_id}} =
+           {:next_level, Campaigns.get_next_level(campaign_progress.level)} do
+      level = campaign_progress.level
+
+      Multi.new()
+      |> Multi.run(:update_campaign_progress, fn _, _ ->
+        if next_level_id == level.id,
+          do: {:ok, nil},
+          else: update_campaign_progress(campaign_progress, next_campaign_id, next_level_id)
+      end)
+      |> Repo.transaction()
+    else
+      {:campaign_data, _transaction_error} -> {:error, :campaign_data_error}
+      {:next_level, _transaction_error} -> {:campaign_progress_error}
+    end
+  end
+
+  def reset_afk_rewards_claim(user_id) do
+    {:ok, user} = get_user(user_id)
+
+    user
+    |> User.changeset(%{last_afk_reward_claim: DateTime.utc_now()})
+    |> Repo.update()
+  end
+
   defp preload(user),
-    do: Repo.preload(user, items: :template, units: [:character, :items], currencies: :currency)
+    do:
+      Repo.preload(user,
+        items: :template,
+        units: [:character, :items],
+        currencies: :currency
+      )
+
+  defp retrieve_campaign_progress_data(user_id, campaign_id) do
+    progress =
+      Repo.get_by(CampaignProgress, user_id: user_id, campaign_id: campaign_id)
+      |> Repo.preload(level: :campaign)
```

Comment:
you already have a getter for this on the `campaigns.ex` module `get_campaign_progress`, maybe adding these preloads there would do it